packages/datadog-instrumentations/src/mongoose.js

```
@@ -26,5 +26,147 @@ addHook({
+            shimmer.wrap(arguments, lastArgumentIndex, originalCb => {
+              return function () {
+                finish()
+
+                return originalCb.apply(this, arguments)
+              }
+            })
+
+            callbackWrapped = true
+          }
+
+          return asyncResource.runInAsyncScope(() => {
+            startCh.publish({
+              filters,
+              methodName
+            })
+
+            const res = method.apply(this, arguments)
+
+            // if it is not callback, wrap exec method and its then
+            if (!callbackWrapped) {
+              shimmer.wrap(res, 'exec', originalExec => {
+                return function wrappedExec () {
+                  const execResult = originalExec.apply(this, arguments)
+
+                  // wrap them method, wrap resolve and reject methods
+                  shimmer.wrap(execResult, 'then', originalThen => {
+                    return function wrappedThen () {
+                      const resolve = arguments[0]
+                      const reject = arguments[1]
+
+                      // not using shimmer here because resolve/reject could be empty
+                      arguments[0] = function wrappedResolve () {
+                        finish()
+
+                        if (resolve) {
+                          return resolve.apply(this, arguments)
+                        }
+                      }
+
+                      arguments[1] = function wrappedReject () {
+                        finish()
+
+                        if (reject) {
+                          return reject.apply(this, arguments)
+                        }
+                      }
+
+                      return originalThen.apply(this, arguments)
+                    }
+                  })
+
+                  return execResult
+                }
+              })
+            }
+            return res
+          })
+        }
+      })
+    } catch (e) {
```

Error swallowing is usually pretty dangerous. What about preceding the wrap line with a conditional:

```js
if (!(methodName in Model)) return
shimmer.wrap(Model, methodName, method => {
```